### PR TITLE
Allow running queries using session properties within a transaction

### PR DESCRIPTION
**Pull request by findinpath**
Resolves: #6642

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/Session.java [hunk 1]
```diff
@@ -389,15 +388,38 @@ public Session beginTransactionId(TransactionId transactionId, TransactionManage
...
-    public Session withDefaultProperties(Map<String, String> systemPropertyDefaults, Map<String, Map<String, String>> catalogPropertyDefaults)
+    public Session withDefaultProperties(Map<String, String> systemPropertyDefaults, Map<String, Map<String, String>> catalogPropertyDefaults, AccessControl accessControl)
     {
         requireNonNull(systemPropertyDefaults, "systemPropertyDefaults is null");
         requireNonNull(catalogPropertyDefaults, "catalogPropertyDefaults is null");
 
-        // to remove this check properties must be authenticated and validated as in beginTransactionId
-        checkState(
-                this.transactionId.isEmpty() && this.connectorProperties.isEmpty(),
-                "Session properties cannot be overridden once a transaction is active");
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/Session.java:400]: In some places we assume session configuration doesn't change during transaction, e.g. in Hive connector.

Please add a test

- setup data with timesamp having nanosecond precision, using timestamp_precision=NANOSECONDS
- for X,Y precisions in (NANOSECONDS,MICROSECONDS,MILLISECONDS) x (NANOSECONDS,MICROSECONDS,MILLISECONDS) (`HiveTimestampPrecision.values()`)
  - in a transaction
    - set timestamp_precision=X
    - read data
    - set timestamp_precision=Y
    - read data

**Review by findepi — CHANGES_REQUESTED** [review]
cc @losipiuk @alexjo2144 @kokosing

File: core/trino-main/src/main/java/io/trino/Session.java (hunk 1 above)
Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/Session.java:400]: @findepi I'm having difficulties to find a way to test the method I have changed.

As can be seen in the image below, the method `io.trino.Session#withDefaultProperties` is being called only from `io.trino.dispatcher.QueuedStatementResource#getStatus(io.trino.spi.QueryId, java.lang.String, long, io.airlift.units.Duration, javax.ws.rs.core.UriInfo, javax.ws.rs.container.AsyncResponse)`

session_withDefaultProperties_call_hierarchy


Can you point me an existing test class in which I could add the test case that you've previously described?

**Comment by sajjoseph** [comment]
This fix is definitely going to help with dbt integration. Some of our end users face this issue while using DBeaver or Pentaho to interact with Trino. Looking forward to see this PR merged into Trino.

Thanks @findinpath for the PR.

**Comment by findepi** [comment]
> This fix is definitely going to help with dbt integration.

@sajjoseph is there a way to avoid creating transactions?

**Comment by sajjoseph** [comment]
@findepi  - I don't think we can avoid transactions.
dbt tool is using transactions while creating table/writing data. Pentaho tool is triggering transactions as well.

**Comment by findepi** [comment]
@sajjoseph i think this is a wrong direction. I understand that it would be awesome if everything was transactional, but reality is -- #6642 is just one of limitations we're going to encounter.
While it's reasonable to remove limitations like this one, it would be even more awesome from user perspective if we make things work end to end.

What would it take to teach dbt to use autocommit?

**Comment by aczajkowski** [comment]
@findinpath I've missed the fact of your PR and implemented suggestion from comment https://github.com/trinodb/trino/blob/ef4b88729d572c25ddeccae420b082cadcfbcf63/core/trino-main/src/main/java/io/trino/Session.java#L403
Sorry for that.
In my PR i've implemented it similar as you + some smaller changes and test coverage
https://github.com/trinodb/trino/pull/9846
How can we combine our work ?

**Comment by findinpath** [comment]
@aczajkowski  no worries. I've briefly looked at your PR and our changes on the `src/main/java` code look mostly the same.
On the other hand, you've added tests to prove the accuracy of your changes. So I tend to believe that my PR is rather obsolete.
 I'd recommend you to reference in a commit comment `Resolves #6642` to help eventually maintainers to understand the background of this change.

**Comment by aczajkowski** [comment]
@findinpath Following @electrum suggestion. Let's maybe cherry pick commits from my PR to yours.
I've made them more granular so you can choose relevant once.
And we will merge your PR then :).

**Comment by findinpath** [comment]
Thank you for the offer.
I think though that your PR is definitely more mature than mine and I'd be only copying stuff from your PR.
I'll carry on  participating in your PR with comments.
I'm fine with closing this PR.

**Pull request closed without merging**